casper/src/test/scala/io/casperlabs/casper/util/CasperUtilTest.scala

```diff
@@ -293,4 +277,88 @@ class CasperUtilTest extends FlatSpec with Matchers with BlockGenerator with Dag
         )
       } yield ()
   }
+
+  // See [[casper/src/test/resources/casper/panoramaForEquivocatorSwimlaneIsEmpty.png]]
+  "panoramaM" should "properly return the panorama of message B, and when V(j)-swimlane is empty or V(j) happens to be an equivocato, put 0 in the corresponding cell." in withStorage {
+    implicit blockStore => implicit blockDagStorage =>
+      val v0 = generateValidator("V0")
+      val v1 = generateValidator("V1")
+
+      val v2                = generateValidator("V2")
+      val v3                = generateValidator("V3")
+      val v4                = generateValidator("V4")
+      val validators        = List(v0, v1, v2, v3, v4)
+      val validatorsToIndex = validators.zipWithIndex.toMap
+
+      val bonds = validators.map(v => Bond(v, 1))
+
+      for {
+        genesis <- createBlock[Task](Seq(), ByteString.EMPTY)
+        b1 <- createBlock[Task](
+               Seq(genesis.blockHash),
+               v0,
+               bonds,
+               Map(v0 -> genesis.blockHash)
+             )
+        b2 <- createBlock[Task](
+               Seq(genesis.blockHash),
+               v3,
+               bonds,
+               Map(v3 -> genesis.blockHash)
+             )
+        b3 <- createBlock[Task](
+               Seq(b1.blockHash),
+               v1,
+               bonds,
+               Map(v0 -> b1.blockHash, v1 -> genesis.blockHash)
+             )
+        b4 <- createBlock[Task](
+               Seq(b1.blockHash),
+               v0,
+               bonds,
+               Map(v0 -> b1.blockHash)
+             )
+        // b5 vote for b2 instead of b1
```

Comment:
```suggestion
        // b5 votes for b2 instead of b1
```
but why? I can see what it does from the code, comments should explain intent.

Reply:
I just wanna show that b7 can see b3 by indirect justifications through `b5`. Other blocks can be seen because of main-parent DAG, just wanna it to be more special.